### PR TITLE
Add a new rule about forbidding sub_folders

### LINKED ISSUE #160: Raise an error when files are in a subfolder
While doing a pass of data quality on https://transport.data.gouv.fr, I noticed a number of GTFS are considered valid despite containing a subfolder:

- https://github.com/etalab/transport-qualite-des-donnees/issues/7 
- https://github.com/etalab/transport-qualite-des-donnees/issues/8
- https://github.com/etalab/transport-qualite-des-donnees/issues/9
- https://github.com/etalab/transport-qualite-des-donnees/issues/10
- https://github.com/etalab/transport-qualite-des-donnees/issues/11
- https://github.com/etalab/transport-qualite-des-donnees/issues/12
- https://github.com/etalab/transport-qualite-des-donnees/issues/13
- https://github.com/etalab/transport-qualite-des-donnees/issues/14

I have opened this issue to discuss this at the spec level:
- https://github.com/google/transit/issues/377

Which raised a return so far:

> We get lots of support requests for exactly this topic in the OpenTripPlanner chat room.
> I would welcome the spec being more clear on this.

The crate used to read GTFS in the validator is smoothly handling that, but other tools may not etc, causing them trouble to ingest the data.

I would personally be happy to consider a file invalid on https://transport.data.gouv.fr if there is a subfolder, since it complicates the ingestion.

Given how the crate works, would it be complicated to return an error in that case (i.e. detect the fact that while data is valid, it wasn't stored at the root)?

To be discussed with the team (@fchabouis @AntoineAugusti), just raising the point with concrete data.

### PR DESCRIPTION
**Pull request by antoine-de**
closes #160

Add an error when data was in a subfolder.

#175 and #174 needs to be merged first

### CODE REVIEW

**Comment by antoine-de** [comment]
@AntoineAugusti maybe some translations also needs to be done for this one.

If you want me to check those translations I can do it.

When all those PR will be merged, I'll also some chores to bump some dependencies.

**Comment by thbar** [comment]
Thanks!

I did a quick check to verify which GTFS may be impacted by this, because we have a global nightly import which raise errors on those cases.

In theory 10 files will be invalidated by this change.

[81291](https://transport.data.gouv.fr/resources/81291)	99100	error	Unzip.Error	File "stops.txt" not present in the zip
[81185](https://transport.data.gouv.fr/resources/81185)	104955	error	Unzip.Error	File "stops.txt" not present in the zip
[81092](https://transport.data.gouv.fr/resources/81092)	106888	error	Unzip.Error	File "stops.txt" not present in the zip
[81090](https://transport.data.gouv.fr/resources/81090)	107110	error	Unzip.Error	File "stops.txt" not present in the zip
[81081](https://transport.data.gouv.fr/resources/81081)	106903	error	Unzip.Error	File "stops.txt" not present in the zip
[81091](https://transport.data.gouv.fr/resources/81091)	106931	error	Unzip.Error	File "stops.txt" not present in the zip
[81021](https://transport.data.gouv.fr/resources/81021)	82763	error	Unzip.Error	File "stops.txt" not present in the zip
[81199](https://transport.data.gouv.fr/resources/81199)	92001	error	Unzip.Error	File "stops.txt" not present in the zip
[81176](https://transport.data.gouv.fr/resources/81176)	88220	error	Unzip.Error	File "stops.txt" not present in the zip

**Comment by antoine-de** [comment]
The first one even have a zip in the subfolder, also containing another GTFS :grinning:

**Comment by AntoineAugusti** [comment]
@antoine-de Could you share a sample JSON payload of the validator with this new rule? To craft great translation messages. https://github.com/etalab/transport-validator/pull/174#issuecomment-1803515440

**Comment by antoine-de** [comment]
Sure, I added a test with a json: https://github.com/etalab/transport-validator/pull/176/commits/60b929b8d1007f67307d82bdce5cfe8f70ad8825

Do you need more than this?

**Comment by AntoineAugusti** [comment]
@antoine-de Awesome, cheers! I was expecting a comment on GitHub but having a test is even better.

I'll craft a PR on the website side to implement a view/messages and link it here.

**Comment by antoine-de** [comment]
I rebased with the last merged PR.

In the meantime, I did a `cargo update` to update all dependencies and fix in the meantime #173

**Comment by AntoineAugusti** [comment]
@antoine-de Just opened a PR on the other repository to deal with this new error, thanks!

**Comment by antoine-de** [comment]
ok, we can merge this after the transport pr is merged, can you approve it when it will be done on your side?

File: src/validators/sub_folder.rs
Comment (AntoineAugusti) [inline src/validators/sub_folder.rs]: Consider removing `dbg!` statements?

File: src/validators/sub_folder.rs
Comment (AntoineAugusti) [inline src/validators/sub_folder.rs]: Remove?

File: src/validators/sub_folder.rs
Comment (AntoineAugusti) [inline src/validators/sub_folder.rs]:
```suggestion
                .details(&format!("Data is contained in a subfolder: {}", parent)),
```

File: src/validators/sub_folder.rs
Comment (AntoineAugusti) [inline src/validators/sub_folder.rs]: Funny to test this only using a single file. I guess that's fine?!

File: src/issues.rs
Comment (AntoineAugusti) [inline src/issues.rs]:
```suggestion
    /// The .txt files within the GTFS are located in a subfolder, which is now explicitly forbidden by the specification
```

File: src/validators/sub_folder.rs
Comment (AntoineAugusti) [inline src/validators/sub_folder.rs]: 😻 🙏

File: src/validators/sub_folder.rs
Comment (AntoineAugusti) [inline src/validators/sub_folder.rs]:
```suggestion
            Some("Data is contained in a subfolder: gtfs".to_string()),
```

**Review by AntoineAugusti — COMMENTED** [review]
LGTM, please have someone else review the Rust part 😉

File: src/validators/sub_folder.rs
Comment (antoine-de) [reply src/validators/sub_folder.rs]: oups :see_no_evil:

File: src/validators/sub_folder.rs
Comment (antoine-de) [reply src/validators/sub_folder.rs]: yep, it can be discussed, but if for example the zip contains a subfolder with metadata or custom file, I did not want to raise an error. So I tested one mandatory file since I think it can cover most usecases.

Do you have a better idea?

Comment (AntoineAugusti) [reply src/validators/sub_folder.rs]: Nope, sounds good. We can adjust if it's a problem. Did you look how MobilityData checks this rule with their validator?

I'm wondering if some GTFS files (GTFS Flex) can have no `stops.txt` file.

Note: there is a `dbg!` statement here left.

Comment (antoine-de) [reply src/validators/sub_folder.rs]: hum, they check all files https://github.com/MobilityData/gtfs-validator/blob/fdd6c9b03bc447664153fac344f63adc0a57a9fe/core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsInput.java#L114-L135

I think even GTFS-FLEX need a stops.txt, but I can check more files if you want (but I'll only be able to do it later this week). 

(oups :see_no_evil:  , sorry for the debug, I removed it)

Comment (AntoineAugusti) [reply src/validators/sub_folder.rs]: A single file should be good, targeting `stops.txt` seems the most obvious

Comment (thbar) [reply src/validators/sub_folder.rs]: Sounds reasonable!

**Review by Tristramg — APPROVED** [review]
Thank you for changing the match!

**Pull request merged**
